source/core_mqtt.c

```diff
@@ -2734,23 +2829,69 @@ MQTTStatus_t MQTT_Connect( MQTTContext_t * pContext,
 
     if( status == MQTTSuccess )
     {
-        /* Resend PUBRELs when reestablishing a session, or clear records for new sessions. */
-        status = handleSessionResumption( pContext, *pSessionPresent );
-    }
+        /* Reset the index and clear the buffer when a new session is established. */
+        pContext->index = 0;
+        ( void ) memset( pContext->networkBuffer.pBuffer, 0, pContext->networkBuffer.size );
+
+        if( *pSessionPresent != true )
+        {
+            /* Clear any existing records if a new session is established. */
+            if( pContext->outgoingPublishRecordMaxCount > 0U )
+            {
+                ( void ) memset( pContext->outgoingPublishRecords,
+                                0x00,
+                                pContext->outgoingPublishRecordMaxCount * sizeof( *pContext->outgoingPublishRecords ) );
+            }
+
+            if( pContext->incomingPublishRecordMaxCount > 0U )
+            {
+                ( void ) memset( pContext->incomingPublishRecords,
+                                0x00,
+                                pContext->incomingPublishRecordMaxCount * sizeof( *pContext->incomingPublishRecords ) );
+            }
+        }
 
-    if( status == MQTTSuccess )
-    {
-        LogInfo( ( "MQTT connection established with the broker." ) );
         pContext->connectStatus = MQTTConnected;
         /* Initialize keep-alive fields after a successful connection. */
         pContext->keepAliveIntervalSec = pConnectInfo->keepAliveSeconds;
         pContext->waitingForPingResp = false;
         pContext->pingReqSendTimeMs = 0U;
     }
+
+    if( stateUpdateHookExecuted == true )
+    {
+        /* Regardless of the status, if the mutex was taken, then it should be relinquished. */
+        MQTT_POST_STATE_UPDATE_HOOK( pContext );
+    }
+
+    if( status == MQTTSuccess && *pSessionPresent == true  )
+    {
+        /* Resend PUBRELs when reestablishing a session */
+        status = handleUncleanSessionResumption( pContext);
+    }
+
```

Comment:
Should we set `pContext->connectStatus = MQTTConnected;` after the above?

Because what happens when a different thread calls `MQTT_CheckConnectStatus`? It might get a MQTTConnected even though the library has not yet handled the session resumption.

I don't think this will violate the specification (see [here](http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/os/mqtt-v3.1.1-os.html#_Toc398718105)). But it seems counter intuitive that while connect has not returned a success status (and may fail during retransmits), a different function can say that the mqtt is connected.

Reply:
The same place to which you linked also says this: _"The Client accepts that any data that it sends before it receives a CONNACK packet from the Server will not be processed if the Server rejects the connection."_.

Is coreMQTT equipped to handle that? What happens in QoS0 packet which the client thinks are delivered to the server but are rejected because the connection was never established. I understand that QoS0 is equivalent of "fire and forget", but since underlying transport is mostly TCP/IP, a user can assume that the QoS0 packet will be delivered to the server and won't get lost. It is a _De Facto_ standard IMO (as opposed to _De Jure_).

Reply:
Ignore the above comment. I see that we indeed wait for the CONNACK